Related

```
[ERROR]   SabreV4ITIPRequestTest>ITIPRequestContract.shouldNotSendUpdateToUninvitedAttendeesWhenOrganizerModifiesOtherInstances:1080 [No new ITIP should be received by uninvited attendee when organizer edits unrelated instance] 
Expected size: 1 but was: 2 in:
["/calendars/68f9dec66a610e73d03c8cae/inbox/sabredav-6cf42637-a5ce-4a79-9d3a-5615da6bd570.ics",
    "/calendars/68f9dec66a610e73d03c8cae/inbox/sabredav-e922ce9b-b608-4c2a-9e68-bb9d91a0074a.ics"]
```

At least now we have a failure to fix!

--- a/lib/CalDAV/Schedule/IMipPlugin.php
+++ b/lib/CalDAV/Schedule/IMipPlugin.php
@@ -101,7 +101,7 @@ class IMipPlugin extends \Sabre\CalDAV\Schedule\IMipPlugin {
             if ($this->testIfEventIsExpired($eventMessage['message'])) {
                 continue;
             }
-             
+
             $message = [
                 'senderEmail' => substr($iTipMessage->sender, 7),
                 'recipientEmail' => substr($iTipMessage->recipient, 7),
@@ -304,6 +304,27 @@ class IMipPlugin extends \Sabre\CalDAV\Schedule\IMipPlugin {
             // Create message if instance have been created or modified
             if (!isset($cancelledInstancesId[$recurrenceId]) && (!isset($previousEventVEvents[$recurrenceId]) ||
                 $this->hasInstanceChanged($previousEventVEvents[$recurrenceId], $currentEventVEvents[$recurrenceId]))) {
+
+                // Check if recipient was attending this occurrence before
+                $isNewOccurrenceException = !isset($previousEventVEvents[$recurrenceId]);
+                $previousVEvent = null;
+
+                if ($isNewOccurrenceException) {
+                    // For new exceptions, check master event to detect removed attendees
+                    $previousVEvent = $previousEventVEvents[self::MASTER_EVENT] ?? null;
+                } else {
+                    // For existing exceptions, check the previous version of this occurrence
+                    $previousVEvent = $previousEventVEvents[$recurrenceId];
+                }
+
+                $wasAttendingBefore = isset($previousVEvent) && $this->isAttending($recipient, $previousVEvent);
+                $isAttendingNow = $this->isAttending($recipient, $currentEventVEvents[$recurrenceId]);
+
+                // Fix for issue #152: Skip AMQP notification if recipient is not and was not attending
+                if (!$isAttendingNow && !$wasAttendingBefore) {
+                    continue;
+                }
+
                 $currentMessage = clone ($scheduledEvent);
                 $modifiedInstance = [];
 
@@ -312,17 +333,18 @@ class IMipPlugin extends \Sabre\CalDAV\Schedule\IMipPlugin {
 
                 $modifiedInstance['message'] = $currentMessage;
 
-                // Check if recipient was attending before
-                // If an exception was created, we check is recipient was attending whole series
-                $previousVEvent = $previousEventVEvents[$recurrenceId] ?? $previousEventVEvents[self::MASTER_EVENT];
-                if (!isset($previousVEvent) || !$this->isAttending($recipient, $previousVEvent)) {
+                // Mark as new event if recipient wasn't attending before
+                if (!$wasAttendingBefore) {
                     $modifiedInstance['newEvent'] = 1;
                 }
 
                 // Can't find the previous vEvent associated with the recurrence id,
                 // which means we're dealing with a newly added recurrence exception here.
                 if (!isset($previousEventVEvents[$recurrenceId])) {
-                    $previousVEvent = $this->getRecurrenceInstance($previousVEvent, $recurrenceId);
+                    $previousVEvent = $previousVEvent ?? $previousEventVEvents[self::MASTER_EVENT] ?? null;
+                    if ($previousVEvent) {
+                        $previousVEvent = $this->getRecurrenceInstance($previousVEvent, $recurrenceId);
+                    }
                 }
 
                 $modifiedInstance['changes'] = $this->getPropertyChanges($previousVEvent, $currentEventVEvents[$recurrenceId]);

--- a/lib/CalDAV/Schedule/Plugin.php
+++ b/lib/CalDAV/Schedule/Plugin.php
@@ -73,6 +73,238 @@ class Plugin extends \Sabre\CalDAV\Schedule\Plugin {
     }
 
     /**
+     * Check if a message should be skipped for an unchanged occurrence
+     *
+     * When modifying one occurrence in a recurring event, SabreDAV's broker creates
+     * messages for ALL occurrences, even those that haven't changed. This method
+     * filters out messages for unchanged occurrences.
+     *
+     * @param ITip\Message $message The iTIP message to check
+     * @param VCalendar $oldObject The old event
+     * @param VCalendar $newObject The new event
+     * @return bool True if the message should be skipped
+     */
+    protected function shouldSkipUnchangedOccurrence(ITip\Message $message, $oldObject, VCalendar $newObject) {
+        // Only apply this filter to REQUEST messages for recurring events
+        if ($message->method !== 'REQUEST') {
+            return false;
+        }
+
+        // Parse oldObject if it's a string (raw iCalendar data)
+        if (is_string($oldObject)) {
+            $oldObject = \Sabre\VObject\Reader::read($oldObject);
+        }
+
+        // Ensure oldObject has VEVENT
+        if (!isset($oldObject->VEVENT)) {
+            return false;
+        }
+
+        // Only apply this filter to recurring events (must have RRULE or RECURRENCE-ID)
+        $hasRecurrence = false;
+        foreach ($oldObject->VEVENT as $vevent) {
+            if (isset($vevent->RRULE) || isset($vevent->{'RECURRENCE-ID'})) {
+                $hasRecurrence = true;
+                break;
+            }
+        }
+        if (!$hasRecurrence) {
+            return false;
+        }
+
+        // Only filter messages with a single VEVENT (single occurrence)
+        // Messages with multiple VEVENTs (bundled occurrences) should not be filtered
+        // as they represent legitimate multi-occurrence invitations
+        $veventCount = count($message->message->VEVENT);
+        if ($veventCount !== 1) {
+            return false;
+        }
+
+        // Get the VEVENT from the message to identify which occurrence this is about
+        $messageEvent = $message->message->VEVENT;
+        if (!$messageEvent) {
+            return false;
+        }
+
+        // Determine the recurrence ID of this message
+        $recurrenceId = isset($messageEvent->{'RECURRENCE-ID'})
+            ? $messageEvent->{'RECURRENCE-ID'}->getValue()
+            : 'master';
+
+        // Find the corresponding VEVENTs in old and new objects
+        $oldVEvent = null;
+        $newVEvent = null;
+
+        foreach ($oldObject->VEVENT as $vevent) {
+            $oldRecurId = isset($vevent->{'RECURRENCE-ID'})
+                ? $vevent->{'RECURRENCE-ID'}->getValue()
+                : 'master';
+            if ($oldRecurId === $recurrenceId) {
+                $oldVEvent = $vevent;
+                break;
+            }
+        }
+
+        foreach ($newObject->VEVENT as $vevent) {
+            $newRecurId = isset($vevent->{'RECURRENCE-ID'})
+                ? $vevent->{'RECURRENCE-ID'}->getValue()
+                : 'master';
+            if ($newRecurId === $recurrenceId) {
+                $newVEvent = $vevent;
+                break;
+            }
+        }
+
+        // If this is a new occurrence (wasn't in oldObject), don't skip
+        if (!$oldVEvent || !$newVEvent) {
+            return false;
+        }
+
+        // Check if recipient was attending this occurrence before
+        $wasAttendingBefore = false;
+        if (isset($oldVEvent->ATTENDEE)) {
+            foreach ($oldVEvent->ATTENDEE as $attendee) {
+                if ($attendee->getNormalizedValue() === $message->recipient) {
+                    $wasAttendingBefore = true;
+                    break;
+                }
+            }
+        }
+
+        // Check if recipient is attending this occurrence now
+        $isAttendingNow = false;
+        if (isset($newVEvent->ATTENDEE)) {
+            foreach ($newVEvent->ATTENDEE as $attendee) {
+                if ($attendee->getNormalizedValue() === $message->recipient) {
+                    $isAttendingNow = true;
+                    break;
+                }
+            }
+        }
+
+        // If recipient wasn't and isn't attending, skip (already handled by broker)
+        // If recipient was attending but isn't now, don't skip (it's a removal)
+        // If recipient wasn't attending but is now, don't skip (it's an addition)
+        // Only skip if recipient was AND is still attending
+        if (!$wasAttendingBefore || !$isAttendingNow) {
+            return false;
+        }
+
+        // Check if the number of occurrences (exceptions) the recipient is invited to has changed
+        // Only count exceptions where the recipient is an attendee
+        $oldExceptionCount = 0;
+        $newExceptionCount = 0;
+        foreach ($oldObject->VEVENT as $vevent) {
+            if (isset($vevent->{'RECURRENCE-ID'})) {
+                // Check if recipient is attending this occurrence
+                if (isset($vevent->ATTENDEE)) {
+                    foreach ($vevent->ATTENDEE as $attendee) {
+                        if ($attendee->getNormalizedValue() === $message->recipient) {
+                            $oldExceptionCount++;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        foreach ($newObject->VEVENT as $vevent) {
+            if (isset($vevent->{'RECURRENCE-ID'})) {
+                // Check if recipient is attending this occurrence
+                if (isset($vevent->ATTENDEE)) {
+                    foreach ($vevent->ATTENDEE as $attendee) {
+                        if ($attendee->getNormalizedValue() === $message->recipient) {
+                            $newExceptionCount++;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        if ($oldExceptionCount !== $newExceptionCount) {
+            return false; // Number of exceptions recipient is invited to changed, don't skip
+        }
+
+        // Check if the occurrence has actually changed
+        // Compare SEQUENCE, DTSTART, DTEND, SUMMARY, LOCATION, DESCRIPTION, etc.
+        $oldSequence = isset($oldVEvent->SEQUENCE) ? $oldVEvent->SEQUENCE->getValue() : 0;
+        $newSequence = isset($newVEvent->SEQUENCE) ? $newVEvent->SEQUENCE->getValue() : 0;
+
+        if ($oldSequence != $newSequence) {
+            return false; // Sequence changed, don't skip
+        }
+
+        // Compare key properties (including EXDATE for occurrence exclusion detection)
+        $properties = ['DTSTART', 'DTEND', 'SUMMARY', 'LOCATION', 'DESCRIPTION', 'STATUS', 'EXDATE'];
+        foreach ($properties as $prop) {
+            $oldValue = isset($oldVEvent->$prop) ? (string)$oldVEvent->$prop : '';
+            $newValue = isset($newVEvent->$prop) ? (string)$newVEvent->$prop : '';
+            if ($oldValue !== $newValue) {
+                return false; // Property changed, don't skip
+            }
+        }
+
+        // Occurrence hasn't changed significantly, skip the message
+        return true;
+    }
+
+    /*
+     * Override to filter IMIP messages for partstat-only changes.
+     * Performance optimization for issue #128:
+     * When an attendee changes their PARTSTAT (accepts/declines), SabreDAV generates
+     * IMIP messages to notify all other attendees. However, attendees don't need to
+     * be notified when another attendee's participation status changes - only the
+     * organizer needs this information.
+     *
+     * This override filters out IMIP messages that:
+     * 1. Have no significant changes (empty $changes)
+     * 2. Are sent to attendees about another attendee's participation change
+     *
+     * As suggested by chibenwa in PR #142 review.
+     */
+    protected function processICalendarChange($oldObject = null, VCalendar $newObject, array $addresses, array $ignore = [], &$modified = false) {
+        $broker = new ITip\Broker();
+        $messages = $broker->parseEvent($newObject, $addresses, $oldObject);
+
+        if ($messages) $modified = true;
+
+        foreach ($messages as $message) {
+            if (in_array($message->recipient, $ignore)) {
+                continue;
+            }
+
+            // Fix for issue #152: Skip delivery for unchanged occurrences
+            // When modifying one occurrence (e.g. creating exception #3), SabreDAV re-processes
+            // all occurrences including unchanged ones (e.g. exception #2). We need to skip
+            // delivering messages for occurrences that haven't actually changed.
+            if ($oldObject && $this->shouldSkipUnchangedOccurrence($message, $oldObject, $newObject)) {
+                continue;
+            }
+
+            $this->deliver($message);
+
+            // Update schedule status for organizer or attendee
+            if (isset($newObject->VEVENT->ORGANIZER) && ($newObject->VEVENT->ORGANIZER->getNormalizedValue() === $message->recipient)) {
+                if ($message->scheduleStatus) {
+                    $newObject->VEVENT->ORGANIZER['SCHEDULE-STATUS'] = $message->getScheduleStatus();
+                }
+                unset($newObject->VEVENT->ORGANIZER['SCHEDULE-FORCE-SEND']);
+            } else {
+                if (isset($newObject->VEVENT->ATTENDEE)) {
+                    foreach ($newObject->VEVENT->ATTENDEE as $attendee) {
+                        if ($attendee->getNormalizedValue() === $message->recipient) {
+                            if ($message->scheduleStatus) {
+                                $attendee['SCHEDULE-STATUS'] = $message->getScheduleStatus();
+                            }
+                            unset($attendee['SCHEDULE-FORCE-SEND']);
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
      *
      * Override default method because:
      *  * user addresses must be the calendar owner ones to handle delegation

--- a/tests/CalDAV/Schedule/IMipPluginRecurrentEventTest.php
+++ b/tests/CalDAV/Schedule/IMipPluginRecurrentEventTest.php
@@ -1196,4 +1196,335 @@ class IMipPluginRecurrentEventTest extends IMipPluginTestBase {
         $plugin->schedule($itipMessage);
         $this->assertEquals('1.1', $itipMessage->scheduleStatus);
     }
+
+    /**
+     * Test for issue #152
+     * When organizer modifies an occurrence that doesn't include an attendee,
+     * that attendee should NOT receive an iTIP notification
+     */
+    function testShouldNotSendUpdateToUninvitedAttendeesWhenOrganizerModifiesOtherInstances()
+    {
+        $plugin = $this->getPlugin();
+
+        // Initial event: recurring event with 3 occurrences
+        // Occurrence #2 has user2 invited
+        // Occurrence #3 does NOT have user2 invited
+        $user1ExistingEvent = join("\r\n", [
+            'BEGIN:VCALENDAR',
+            'VERSION:2.0',
+            'PRODID:-//Sabre//Sabre VObject 4.1.3//EN',
+            'CALSCALE:GREGORIAN',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RRULE:FREQ=DAILY;COUNT=3',
+            'DTSTART:20201028T170000Z',
+            'DTEND:20201028T173000Z',
+            'SUMMARY:Test',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'DTSTAMP:20201025T145516Z',
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RECURRENCE-ID:20201029T170000Z',
+            'DTSTART:20201029T170000Z',
+            'DTEND:20201029T173000Z',
+            'SUMMARY:Test - Instance #2 (user2 invited)',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user2Email,
+            'DTSTAMP:20201025T145516Z',
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RECURRENCE-ID:20201030T170000Z',
+            'DTSTART:20201030T170000Z',
+            'DTEND:20201030T173000Z',
+            'SUMMARY:Test - Instance #3 (user2 NOT invited)',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'DTSTAMP:20201025T145516Z',
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'END:VCALENDAR'
+        ]);
+
+        $plugin->setFormerEventICal($user1ExistingEvent);
+        $plugin->setNewEvent(false);
+
+        // Organizer modifies occurrence #3 (where user2 is NOT invited)
+        $scheduledIcal = join("\r\n", [
+            'BEGIN:VCALENDAR',
+            'VERSION:2.0',
+            'PRODID:-//Sabre//Sabre VObject 4.1.3//EN',
+            'CALSCALE:GREGORIAN',
+            'METHOD:REQUEST',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RRULE:FREQ=DAILY;COUNT=3',
+            'DTSTART:20201028T170000Z',
+            'DTEND:20201028T173000Z',
+            'SUMMARY:Test',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'DTSTAMP:20201025T145516Z',
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RECURRENCE-ID:20201029T170000Z',
+            'DTSTART:20201029T170000Z',
+            'DTEND:20201029T173000Z',
+            'SUMMARY:Test - Instance #2 (user2 invited)',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user2Email,
+            'DTSTAMP:20201025T145516Z',
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RECURRENCE-ID:20201030T170000Z',
+            'DTSTART:20201030T180000Z',
+            'DTEND:20201030T183000Z',
+            'SUMMARY:Test - Instance #3 MODIFIED (user2 NOT invited)',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'DTSTAMP:20201027T182723Z',
+            'SEQUENCE:1',
+            'END:VEVENT',
+            'END:VCALENDAR',
+            ''
+        ]);
+
+        $itipMessage = new \Sabre\VObject\ITip\Message();
+        $itipMessage->uid = 'ab9e450a-3080-4274-affd-fdd0e9eefdcc';
+        $itipMessage->component = 'VEVENT';
+        $itipMessage->method = 'REQUEST';
+        $itipMessage->sequence = 1;
+        $itipMessage->sender = 'mailto:' . $this->user1Email;
+        $itipMessage->recipient = 'mailto:' . $this->user2Email;
+        $itipMessage->scheduleStatus = null;
+        $itipMessage->significantChange = true;
+        $itipMessage->hasChange = true;
+        $itipMessage->message = Reader::read($scheduledIcal);
+
+        // EXPECTED: No notification should be sent to user2
+        // because the modified occurrence (#3) does not include user2 in ATTENDEE list
+        $this->amqpPublisher->expects($this->never())
+            ->method('publish');
+
+        $plugin->schedule($itipMessage);
+        $this->assertEquals('1.1', $itipMessage->scheduleStatus);
+    }
+
+    /**
+     * Test for issue #152 - complementary test
+     * When organizer modifies an occurrence to invite an attendee,
+     * that attendee SHOULD receive an iTIP notification
+     */
+    function testShouldSendRequestWhenOrganizerInvitesAttendeeToSpecificOccurrence()
+    {
+        $plugin = $this->getPlugin();
+
+        // Initial event: recurring event where user2 is NOT invited to any occurrence
+        $user1ExistingEvent = join("\r\n", [
+            'BEGIN:VCALENDAR',
+            'VERSION:2.0',
+            'PRODID:-//Sabre//Sabre VObject 4.1.3//EN',
+            'CALSCALE:GREGORIAN',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RRULE:FREQ=DAILY;COUNT=3',
+            'DTSTART:20501028T170000Z',
+            'DTEND:20501028T173000Z',
+            'SUMMARY:Test',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'DTSTAMP:20501025T145516Z',
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'END:VCALENDAR'
+        ]);
+
+        $plugin->setFormerEventICal($user1ExistingEvent);
+        $plugin->setNewEvent(false);
+
+        // Organizer modifies occurrence #2 to invite user2
+        $scheduledIcal = join("\r\n", [
+            'BEGIN:VCALENDAR',
+            'VERSION:2.0',
+            'PRODID:-//Sabre//Sabre VObject 4.1.3//EN',
+            'CALSCALE:GREGORIAN',
+            'METHOD:REQUEST',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RRULE:FREQ=DAILY;COUNT=3',
+            'DTSTART:20501028T170000Z',
+            'DTEND:20501028T173000Z',
+            'SUMMARY:Test',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'DTSTAMP:20501025T145516Z',
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'BEGIN:VEVENT',
+            'UID:ab9e450a-3080-4274-affd-fdd0e9eefdcc',
+            'RECURRENCE-ID:20501029T170000Z',
+            'DTSTART:20501029T170000Z',
+            'DTEND:20501029T173000Z',
+            'SUMMARY:Test - Instance #2 (user2 NOW invited)',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user2Email,
+            'DTSTAMP:20501027T182723Z',
+            'SEQUENCE:1',
+            'END:VEVENT',
+            'END:VCALENDAR',
+            ''
+        ]);
+
+        $itipMessage = new \Sabre\VObject\ITip\Message();
+        $itipMessage->uid = 'ab9e450a-3080-4274-affd-fdd0e9eefdcc';
+        $itipMessage->component = 'VEVENT';
+        $itipMessage->method = 'REQUEST';
+        $itipMessage->sequence = 1;
+        $itipMessage->sender = 'mailto:' . $this->user1Email;
+        $itipMessage->recipient = 'mailto:' . $this->user2Email;
+        $itipMessage->scheduleStatus = null;
+        $itipMessage->significantChange = true;
+        $itipMessage->hasChange = true;
+        $itipMessage->message = Reader::read($scheduledIcal);
+
+        // EXPECTED: user2 SHOULD receive an iTIP notification
+        // because the organizer is inviting user2 to occurrence #2
+        $this->amqpPublisher->expects($this->once())
+            ->method('publish')
+            ->with(
+                $this->equalTo('calendar:event:notificationEmail:send'),
+                $this->callback(function ($message) {
+                    $decoded = json_decode($message, true);
+                    return $decoded['recipientEmail'] === 'rudyvoller@om.com'
+                        && $decoded['method'] === 'REQUEST'
+                        && isset($decoded['isNewEvent'])
+                        && $decoded['isNewEvent'] === true;
+                })
+            );
+
+        $plugin->schedule($itipMessage);
+        $this->assertEquals('1.1', $itipMessage->scheduleStatus);
+    }
+
+    /**
+     * Test for issue #152: Organizer modifies an occurrence that attendee is not invited to
+     *
+     * Scenario:
+     * 1. Bob creates recurring event (3 days) with only himself
+     * 2. Bob invites Cedric to occurrence #2 only
+     * 3. Bob creates occurrence #3 (sends complete document with unchanged occurrence #2)
+     * 4. Cedric should NOT receive notification (not for unchanged #2, not for new #3)
+     */
+    function testShouldNotSendUpdateToUninvitedAttendeeWhenOrganizerModifiesOtherOccurrence()
+    {
+        $plugin = $this->getPlugin();
+        $plugin->setNewEvent(false);
+
+        // STEP 1: Bob had previously invited Cedric to occurrence #2 only
+        $formerEventIcal = join("\r\n", [
+            'BEGIN:VCALENDAR',
+            'VERSION:2.0',
+            'PRODID:-//Sabre//Sabre VObject 4.1.3//EN',
+            'CALSCALE:GREGORIAN',
+            'BEGIN:VEVENT',
+            'UID:test-event-152',
+            'DTSTART:20260322T090000Z',
+            'DTEND:20260322T100000Z',
+            'RRULE:FREQ=DAILY;COUNT=3',
+            'SUMMARY:Daily meeting',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'BEGIN:VEVENT',
+            'UID:test-event-152',
+            'DTSTART:20260323T090000Z',
+            'DTEND:20260323T100000Z',
+            'RECURRENCE-ID:20260323T090000Z',
+            'SUMMARY:Daily meeting',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user2Email,
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'END:VCALENDAR'
+        ]);
+
+        $plugin->setFormerEventICal($formerEventIcal);
+
+        // STEP 2: Bob creates occurrence #3 (sends complete VCALENDAR with unchanged occurrence #2)
+        // This simulates a PUT with the complete event containing all occurrences
+        $scheduledIcal = join("\r\n", [
+            'BEGIN:VCALENDAR',
+            'VERSION:2.0',
+            'PRODID:-//Sabre//Sabre VObject 4.1.3//EN',
+            'CALSCALE:GREGORIAN',
+            'METHOD:REQUEST',
+            'BEGIN:VEVENT',
+            'UID:test-event-152',
+            'DTSTART:20260322T090000Z',
+            'DTEND:20260322T100000Z',
+            'RRULE:FREQ=DAILY;COUNT=3',
+            'SUMMARY:Daily meeting',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'BEGIN:VEVENT',
+            'UID:test-event-152',
+            'DTSTART:20260323T090000Z',
+            'DTEND:20260323T100000Z',
+            'RECURRENCE-ID:20260323T090000Z',
+            'SUMMARY:Daily meeting',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user2Email,
+            'SEQUENCE:0',
+            'END:VEVENT',
+            'BEGIN:VEVENT',
+            'UID:test-event-152',
+            'DTSTART:20260324T090000Z',
+            'DTEND:20260324T100000Z',
+            'RECURRENCE-ID:20260324T090000Z',
+            'SUMMARY:Updated instance (Day 3)',
+            'ORGANIZER:mailto:' . $this->user1Email,
+            'ATTENDEE:mailto:' . $this->user1Email,
+            'SEQUENCE:2',
+            'END:VEVENT',
+            'END:VCALENDAR',
+            ''
+        ]);
+
+        $itipMessage = new \Sabre\VObject\ITip\Message();
+        $itipMessage->uid = 'test-event-152';
+        $itipMessage->component = 'VEVENT';
+        $itipMessage->method = 'REQUEST';
+        $itipMessage->sequence = 2;
+        $itipMessage->sender = 'mailto:' . $this->user1Email;
+        $itipMessage->recipient = 'mailto:' . $this->user2Email;
+        $itipMessage->scheduleStatus = 'null';
+        $itipMessage->significantChange = true;
+        $itipMessage->hasChange = true;
+        $itipMessage->message = Reader::read($scheduledIcal);
+
+        // EXPECTED: user2 (Cedric) should NOT receive any notification because:
+        // - Occurrence #2: unchanged (same SEQUENCE, same properties) - SKIP
+        // - Occurrence #3: Cedric not invited - SKIP
+        $this->amqpPublisher->expects($this->never())
+            ->method('publish');
+
+        $plugin->schedule($itipMessage);
+        $this->assertEquals('1.1', $itipMessage->scheduleStatus);
+    }
 }

